File: backend/src/main/scala/domains/user/http/api/UpdateUserPreferences.scala
package domains.user.http.api

import cats.effect.IO
import domains.auth.http.AuthenticatedApi
import domains.auth.objects.AuthUser
import domains.user.http.UserApiSupport
import domains.user.http.codec.UserHttpCodecs.given
import domains.user.objects.Username
import domains.user.objects.request.UpdateOwnPreferencesRequest
import domains.user.objects.response.UserSettingsResponse
import domains.user.table.user.UserTable
import io.circe.Encoder
import org.http4s.circe.CirceEntityCodec.*
import org.http4s.{Method, Request, Status}
import shared.http.{ApiMessages, ApiPath, HttpApiError, PathParams}

import java.sql.Connection

object UpdateUserPreferences extends AuthenticatedApi[(Username, UpdateOwnPreferencesRequest), UserSettingsResponse]:

  override val method: Method = Method.POST
  override val path: ApiPath = ApiPath("/api/users/:targetUsername/settings/preferences")
  override val successStatus: Status = Status.Ok
  override protected val outputEncoder: Encoder[UserSettingsResponse] = summon[Encoder[UserSettingsResponse]]

  override def decode(request: Request[IO], pathParams: PathParams): IO[(Username, UpdateOwnPreferencesRequest)] =
    for
      rawUsername <- HttpApiError.fromEitherBadRequest(pathParams.require("targetUsername"))
      updateRequest <- request.as[UpdateOwnPreferencesRequest]
    yield (Username.canonical(rawUsername), updateRequest)

  override def plan(
    connection: Connection,
    actor: AuthUser,
    input: (Username, UpdateOwnPreferencesRequest)
  ): IO[UserSettingsResponse] =
    val (targetUsername, request) = input
    for
      _ <- HttpApiError.ensure(
        targetUsername.value == actor.username.value || actor.siteManager,
        HttpApiError.forbidden(ApiMessages.siteManagerRequired)
      )
      maybeTargetUser <- UserTable.findByUsername(connection, targetUsername)
      targetUser <- maybeTargetUser match
        case Some(user) => IO.pure(user)
        case None => HttpApiError.raise(HttpApiError.notFound(ApiMessages.userNotFound))
      updated <- UserTable.updateSettings(
        connection,
        targetUser.username,
        displayName = targetUser.displayName,
        displayMode = request.preferences.displayMode,
        locale = request.preferences.locale,
        problemTitleDisplayMode = request.preferences.problemTitleDisplayMode,
        autoMarkMessageRead = request.preferences.autoMarkMessageRead
      )
      user <- updated match
        case Some(user) => IO.pure(user)
        case None => HttpApiError.raise(HttpApiError.notFound(ApiMessages.userNotFound))
    yield UserApiSupport.toUserSettingsResponse(user)

File: frontend/src/apis/user/UpdateUserPreferences.ts
import type { APIWithSessionMessage } from '@/system/api/api-message'
import type { SessionResponse } from '@/objects/auth/response/SessionResponse'
import type { UpdateManagedUserPreferencesRequest } from '@/objects/user/request/UpdateManagedUserPreferencesRequest'
import type { UpdateOwnPreferencesRequest } from '@/objects/user/request/UpdateOwnPreferencesRequest'
import type { Username } from '@/objects/user/Username'
import { usernameValue } from '@/objects/user/Username'

/** 更新用户偏好的请求联合；支持用户自助和管理员代管两种请求来源。 */
type UpdateUserPreferencesRequest = UpdateOwnPreferencesRequest | UpdateManagedUserPreferencesRequest

/** 更新指定用户偏好；输入用户名和偏好请求，输出新的会话设置快照。 */
export class UpdateUserPreferences implements APIWithSessionMessage<SessionResponse> {
  declare readonly responseType?: SessionResponse
  readonly method = 'POST'
  readonly apiPath: string
  private readonly request: UpdateUserPreferencesRequest

  constructor(username: Username, request: UpdateUserPreferencesRequest) {
    this.apiPath = `users/${encodeURIComponent(usernameValue(username))}/settings/preferences`
    this.request = request
  }

  body(): UpdateUserPreferencesRequest {
    return this.request
  }
}
